### PR TITLE
[blazor] Use JSImport for lazy loading assemblies

### PR DESCRIPTION
**Pull request by pavelsavara**
## Motivation

- The current `getLazyAssemblies` is implemented using legacy/obsolete JS interop.  
- legacy interop has problems with multi-threading and uses unsupported API.
- legacy interop is not CSP compliant

Contributes to https://github.com/dotnet/aspnetcore/issues/37787

## Description
- use new generated JavaScript interop with `[JSImport]`
- moved async parallel download to C# side
- only one allocation & copy of the DLL bytes from fetch buffer directly to wasm memory.
- the new implementation is loading DLLs asynchronously as they arrive, which should be faster overall

### CODE REVIEW

**Comment by pavelsavara** [comment]
CI `Could not connect to security.ubuntu.com:80 (185.125.190.36). - connect (111: Connection refused) ` is unrelated

**Review by maraf — APPROVED** [review]
LGTM 👍

**Review by ilonatommy — COMMENTED** [review]
I like it, the new version is easier to read. The only idea I got was: can we have names in resources in plural form? `resources.lazyAssembly`, `resources.pdb`, `resources.assembly` -> `resources.lazyAssemblies`, `resources.pdbs`, `resources.assemblies`

**Comment by pavelsavara** [comment]
> I like it, the new version is easier to read. The only idea I got was: can we have names in resources in plural form? `resources.lazyAssembly`, `resources.pdb`, `resources.assembly` -> `resources.lazyAssemblies`, `resources.pdbs`, `resources.assemblies`

Those are from `blazor.boot.json` and I think we may need to keep it backward compatible.

File: src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs [hunk 1]
```diff
@@ -104,39 +102,34 @@ private async Task<IEnumerable<Assembly>> LoadAssembliesInClientAsync(IEnumerabl
...
-        var loadedAssemblies = new List<Assembly>();
-#pragma warning disable CS0618 // Type or member is obsolete
-        var assemblies = jsRuntime.InvokeUnmarshalled<byte[][]>(ReadLazyAssemblies);
-        var pdbs = jsRuntime.InvokeUnmarshalled<byte[][]>(ReadLazyPDBs);
-#pragma warning restore CS0618 // Type or member is obsolete
+    [RequiresUnreferencedCode("Types and members the loaded assemblies depend on might be removed")]
+    [SupportedOSPlatform("browser")]
+    private async Task LoadAssembly(string assemblyToLoad, List<Assembly> loadedAssemblies)
+    {
+        JSObject files = await LazyAssemblyLoaderInterop.LoadLazyAssembly(assemblyToLoad);
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs:116]: Nit:
```suggestion
        using var files = await LazyAssemblyLoaderInterop.LoadLazyAssembly(assemblyToLoad);
```

File: src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs [hunk 2]
```diff
@@ -104,39 +102,34 @@ private async Task<IEnumerable<Assembly>> LoadAssembliesInClientAsync(IEnumerabl
...
+        byte[]? pdbBytes = files.GetPropertyAsByteArray("pdb");
+        Assembly loadedAssembly = pdbBytes == null
+                    ? AssemblyLoadContext.Default.LoadFromStream(new MemoryStream(dllBytes))
+                    : AssemblyLoadContext.Default.LoadFromStream(new MemoryStream(dllBytes), new MemoryStream(pdbBytes));
 
-        return loadedAssemblies;
+        loadedAssemblies.Add(loadedAssembly);
+        _loadedAssemblyCache!.Add(assemblyToLoad);
+
+        files.Dispose();
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs:127]: Nit (if earlier `using var` suggestion is applied):
```suggestion
```

File: src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs [hunk 3]
```diff
@@ -104,39 +102,34 @@ private async Task<IEnumerable<Assembly>> LoadAssembliesInClientAsync(IEnumerabl
...
-            var pdb = pdbs[i];
-            var loadedAssembly = pdb.Length == 0 ?
-                AssemblyLoadContext.Default.LoadFromStream(new MemoryStream(assembly)) :
-                AssemblyLoadContext.Default.LoadFromStream(new MemoryStream(assembly), new MemoryStream(pdb));
-            loadedAssemblies.Add(loadedAssembly);
-            _loadedAssemblyCache.Add(loadedAssembly.GetName().Name + ".dll");
-        }
+        byte[] dllBytes = files.GetPropertyAsByteArray("dll")!;
+        byte[]? pdbBytes = files.GetPropertyAsByteArray("pdb");
+        Assembly loadedAssembly = pdbBytes == null
```
Comment (MackinnonBuck) [inline src/Components/WebAssembly/WebAssembly/src/Services/LazyAssemblyLoader.cs:120]: Nit: `var` for each of these

File: src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts [hunk 4]
```diff
@@ -144,13 +144,13 @@ export const monoPlatform: Platform = {
     return ((baseAddress as any as number) + (fieldOffset || 0)) as any as T;
   },
 
-  beginHeapLock: function() {
+  beginHeapLock: function () {
```
Comment (MackinnonBuck) [inline src/Components/Web.JS/src/Platform/Mono/MonoPlatform.ts:147]: Was this intentional? `.eslintrc.js` is configured to treat this as a styling error.

**Review by MackinnonBuck — APPROVED** [review]
LGTM! Just added some nits, feel free to ignore if you disagree with the changes.

**Pull request merged**